external/dhcpd/dhcpd_lwip.c

```diff
@@ -92,15 +87,85 @@
 #define DHCP_SERVER_PORT         67
 #define DHCP_CLIENT_PORT         68
 
-/****************************************************************************
- * Private Types
- ****************************************************************************/
-/****************************************************************************
- * Private Data
- ****************************************************************************/
+static pthread_t g_dhcpd_tid = -1;
+static int g_dhcpd_term = 0;
+
+struct dhcp_join_data {
+	char *intf;
+	dhcp_sta_joined fn;
+};
+typedef struct dhcp_join_data dhcp_join_data_s;
+
 /****************************************************************************
  * Private Functions
  ****************************************************************************/
+#define DHCPD_MQ_NAME "dhcpd_queue"
+#define DHCPD_MQ_LEN 11
+#define DHCPD_MQ_MAX_LEN 20
+
+void *_dhcpd_join_handler(void *arg)
+{
+	dhcp_join_data_s *data = (dhcp_join_data_s *)arg;
+	int sockfd = 0;
+	struct req_lwip_data req;
+	int ret = ERROR;
+	struct mq_attr attr;
+	attr.mq_maxmsg = DHCPD_MQ_MAX_LEN;
+	attr.mq_msgsize = DHCPD_MQ_LEN;
+	attr.mq_flags = 0;
+	attr.mq_curmsgs = 0;
+
+	mqd_t md = mq_open(DHCPD_MQ_NAME, O_RDWR | O_NONBLOCK |O_CREAT, 0666, &attr);
+	if (md == (mqd_t)ERROR) {
+		ndbg("Failed to open mq\n");
+		goto go_out;
+	}
+
+	sockfd = socket(AF_INET, SOCK_DGRAM, 0);
+	if (sockfd < 0) {
+		ndbg("socket() failed with errno: %d\n", errno);
+		goto go_out;
+	}
+
+	memset(&req, 0, sizeof(req));
+	req.type = DHCPDSTART;
+	req.host_name = data->intf;
+
+	ret = ioctl(sockfd, SIOCLWIP, (unsigned long)&req);
+	if (ret == ERROR) {
+		ndbg("ioctl() failed with errno: %d\n", errno);
+		goto go_out;
+	}
+
+	while (!g_dhcpd_term) {
+		char msg[DHCPD_MQ_LEN];
+		memset(msg, 0, DHCPD_MQ_LEN);
+		int prio = 0;
+		int nbytes = mq_receive(md, msg, DHCPD_MQ_LEN, &prio);
+		if (nbytes <= 0) {
+			if (errno != EAGAIN) {
+				ndbg("mq receive none (errno %d)\n", errno);
+				break;
+			}
+		} else {
```

Comment:
This is not good, why don't you define message and check returned message?
I think Callback mechanism is much more better than your current logic.....

Reply:
I agree with you, but we want to separate kernel and user space strictly, so callback might not be appropriate in this logic. I just added a logic to check whether the received msg is proper or not. Thanks for the comment.